cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/kafka/KafkaSimpleApiConsumer.java

```diff
@@ -0,0 +1,415 @@
+import co.cask.cdap.templates.etl.api.realtime.SourceState;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.Service;
+
+import org.apache.twill.kafka.client.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * <p>
+ *   Wrapper class for {@code Apache Kafka} consumer using the simple APIs to manage the last offset for a particular
+ *   partition of a topic.
+ *
+ *   Most of the code are borrowed from the {@code cdap-kafka-pack} project and modified to be suit the stream source.
+ * </p>
+ *
+ * @param <KEY> Type of message key
+ * @param <PAYLOAD> Type of message value
+ * @param <OFFSET> Type of offset object
+ */
+public abstract class KafkaSimpleApiConsumer<KEY, PAYLOAD, OFFSET> {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaSimpleApiConsumer.class);
+
+  protected static final int SO_TIMEOUT = 5 * 1000;           // 5 seconds.
+
+  private final Function<KafkaConsumerInfo<OFFSET>, OFFSET> consumerToOffset =
+    new Function<KafkaConsumerInfo<OFFSET>, OFFSET>() {
+      @Override
+      public OFFSET apply(KafkaConsumerInfo<OFFSET> input) {
+        return input.getReadOffset();
+      }
+    };
+
+  private Function<ByteBuffer, KEY> keyDecoder;
+  private Function<ByteBuffer, PAYLOAD> payloadDecoder;
+  private KafkaConfig kafkaConfig;
+  private Map<TopicPartition, KafkaConsumerInfo<OFFSET>> consumerInfos;
+  private Map<String, byte[]> offsetStore;
+
+  private RealtimeContext sourceContext;
+
+  /**
+   * <p>
+   *   The method should be called to initialized the consumer when being used by the {@code RealtimeSource}
+   * </p>
+   *
+   * @param context
+   * @throws Exception
+   */
+  public void initialize(RealtimeContext context) throws Exception {
+    sourceContext = context;
+
+    // Tries to detect key and payload decoder based on the class parameterized type.
+    Type superType = TypeToken.of(getClass()).getSupertype(KafkaSimpleApiConsumer.class).getType();
+
+    // Tries to detect Key and Payload type for creating decoder for them
+    if (superType instanceof ParameterizedType) {
+      // Extract Key and Payload types
+      Type[] typeArgs = ((ParameterizedType) superType).getActualTypeArguments();
+
+      keyDecoder = createKeyDecoder(typeArgs[0]);
+      payloadDecoder = createPayloadDecoder(typeArgs[1]);
+    }
+
+    // Configure kafka
+    DefaultKafkaConfigurer kafkaConfigurer = new DefaultKafkaConfigurer();
+    configureKafka(kafkaConfigurer);
+
+    if (kafkaConfigurer.getZookeeper() == null && kafkaConfigurer.getBrokers() == null) {
```

Comment:
ah! looks like you catch that case I mentioned for validation earlier here. Hmmm, I think you can do this much earlier, in the ``configurePipeline()`` method, so this won't be needed.